M1FExplained/Chapter25/Exercises/Exercise01.lean

```diff
@@ -0,0 +1,197 @@
+  by_contra h
+  have h₁ : (a.1 + 1) * (b.1 + 1) = 0 :=
+    calc
+      (a.1 + 1) * (b.1 + 1) = a.1 * b.1 + a.1 + b.1 + 1 := by ring
+      _ = 0 := by
+        rw [h]
+        norm_num
+  have h₂ : a.1 + 1 = 0 ∨ b.1 + 1 = 0 := Iff.mp zero_eq_mul (id (Eq.symm h₁))
+  have h₃ : a.1 = -1 ∨ b.1 = -1 := by 
+    rcases h₂ with (h | h₁)
+    left
+    exact Iff.mp add_eq_zero_iff_eq_neg h
+    right
+    exact Iff.mp add_eq_zero_iff_eq_neg h₁
+  unfold real_not_minus_1 at a b
+  rcases h₃ with (h₄|h₅)
+  have := a.2
+  exact this h₄
+  have := b.2
+  exact this h₅
+
+@[simp,norm_cast]
+lemma coe_zero : ((0 : real_not_minus_1) : ℝ) = 0 := by rfl 
+
+instance : Add real_not_minus_1 where
+  add a b :=  ⟨a.1 * b.1 + a.1 + b.1, add_closed2 a b⟩
+
+@[norm_cast]
+lemma rnm1_coe_add (a b : real_not_minus_1) : (a + b : real_not_minus_1) = (a : ℝ) * b + a + b := by rfl
+
+lemma add_assoc2 (a b c : real_not_minus_1) : a + b + c = a + (b + c) := by
+  ext
+  push_cast
+  ring
+
+lemma neg_real_neq_one (a : real_not_minus_1) : -a.1/(a.1 + 1) ≠ -1 := by
+  intro h
+  simp at h
+  have : a.1 + 1 ≠ 0 := by
+    have := a.2
+    intro h₀
+    apply this
+    exact Iff.mp add_eq_zero_iff_eq_neg h₀
+  have h₁ : a.1/(a.1 + 1) = 1  := by
+    calc
+      a.1/ (a.1 + 1) = -(-a.1/(a.1 + 1)) := by ring
+      _ = -(-1) := by rw [h]
+    ring
+  have h₂ : a.1 = a.1 + 1 := by
+    calc
+      a.1 = a.1/(a.1 + 1) * (a.1 + 1) := Iff.mp (div_eq_iff this) rfl
+      _ = a.1 + 1 := by
+        rw [h₁]
+        ring
+  simp at h₂
+
+noncomputable instance : Neg real_not_minus_1 where
+  neg a := ⟨-a/(a + 1), neg_real_neq_one a⟩
+
+@[norm_cast]
+lemma coe_neg (a : real_not_minus_1) : (-a : real_not_minus_1) = -(a : ℝ)/(a+1) := by rfl
+
+noncomputable instance : AddGroup real_not_minus_1 where
+  add_assoc := add_assoc2
+  zero_add := by
+    intro a
+    ext
+    push_cast
+    norm_num
+  add_zero := by
+    intro a
+    ext
+    push_cast
+    norm_num
+  add_left_neg := by
+    intro a
+    unfold real_not_minus_1
+    ext
+    push_cast
+    calc
+      _ = -a.1/(a.1+1) *a.1 + -a.1/(a.1 + 1) + (a.1 * (a.1 +1))/(a.1 + 1) := by
+        simp
+        have : a.1 + 1 ≠ 0 := by
+          have := a.2
+          intro h₀
+          apply this
+          exact Iff.mp add_eq_zero_iff_eq_neg h₀
+        exact Iff.mpr (eq_div_iff this) rfl
+      _ = 0 := by ring
+
+-- part iv
+
+abbrev part_iv :=
+{ x : ℝ // x ≥ 0}
+
+
+instance zero_part_iv : Zero part_iv := ⟨0,  by norm_num⟩ 
+
+lemma add_close_part_iv (a b : part_iv) : max a.1 b.1 ≥ 0 := by
+  rcases a with ⟨a,ha⟩
+  rcases b with ⟨b,hb⟩
+  simp
+  exact Or.inr hb
+
+noncomputable instance add_part_iv : Add part_iv where
+  add a b:= ⟨max a.1 b.1, add_close_part_iv a b⟩
+
+@[simp]
+lemma add_part_iv_coe (a b : part_iv): (a + b : part_iv) = max (a : ℝ) b := by rfl 
+
+example (a b : part_iv) (h : a.1 > 0): a.1 + b.1 ≠ 0 := by
+  intro h₀
+  norm_cast at h₀
+  simp at h₀
```

Comment:
`simp` should only be used to close goals. Please replace with the output of `simp?`